src/Features/Core/Portable/Completion/CompletionContext.cs

```diff
@@ -0,0 +1,174 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Completion
+{
+    /// <summary>
+    /// The context presented to a <see cref="CompletionProvider"/> when providing completions.
+    /// </summary>
+    public sealed class CompletionContext
+    {
+        private readonly List<CompletionItem> _items;
+
+        internal IReadOnlyList<CompletionItem> Items
+        {
+            get { return _items; }
+        }
+
+        internal CompletionProvider Provider { get; }
+
+        /// <summary>
+        /// The document that completion was invoked within.
+        /// </summary>
+        public Document Document { get; }
+
+        /// <summary>
+        /// The caret position when completion was triggered.
+        /// </summary>
+        public int Position { get; }
+
+        /// <summary>
+        /// The span of the syntax element at the caret position.
+        /// 
+        /// This is the most common value used for <see cref="CompletionItem.Span"/> and will
+        /// be automatically assigned to any <see cref="CompletionItem"/> that has no <see cref="CompletionItem.Span"/> specified.
+        /// </summary>
+        public TextSpan DefaultItemSpan { get; }
+
+        /// <summary>
+        /// The triggering action that caused completion to be started.
+        /// </summary>
+        public CompletionTrigger Trigger { get; }
+
+        /// <summary>
+        /// The options that completion was started with.
+        /// </summary>
+        public OptionSet Options { get; }
+
+        /// <summary>
+        /// The cancellation token to use for this operation.
+        /// </summary>
+        public CancellationToken CancellationToken { get; }
+
+        /// <summary>
+        /// Set to true if the items added here should be the only items presented to the user.
+        /// </summary>
+        public bool IsExclusive { get; set; }
+
+        /// <summary>
+        /// Creates a <see cref="CompletionContext"/> instance.
+        /// </summary>
+        public CompletionContext(
+            CompletionProvider provider,
+            Document document,
+            int position,
+            TextSpan defaultSpan,
+            CompletionTrigger trigger,
+            OptionSet options,
+            CancellationToken cancellationToken)
+        {
```

Comment:
Should this constructor be public?  Generally i think of the contexts as being created only by the controller infrastructure, and then passed to clients.  We don't ever expect clients to create them themselves.